### PR TITLE
Implement EditCardServlet & TranslationServlet with tests, merge tdd7 to tdd6

### PR DESCRIPTION
**Pull request by ethyus**
Screenshot 2020-08-13 at 11 19 56 PM
Screenshot 2020-08-13 at 11 20 40 PM

### CODE REVIEW

File: src/test/java/com/google/sps/servlets/EditCardServletTest.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Entity;
+
+import com.google.gson.Gson;
+import com.google.common.collect.ImmutableMap;
+import com.google.sps.data.Card;
+import com.google.sps.tool.EntityTestingTool;
+import java.util.List; 
+import java.util.ArrayList;
```
Comment (Froggerson) [inline src/test/java/com/google/sps/servlets/EditCardServletTest.java]: Is there a need to import arraylists, hashmaps, maps, and lists here?

File: src/main/java/com/google/sps/servlets/EditCardServlet.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    // User submitted form without selecting a file, so we can't get a BlobKey. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so the blobInfo has 0 byte (live server)
```
Comment (Froggerson) [inline src/main/java/com/google/sps/servlets/EditCardServlet.java:110]: Nice comments!

**Review by Froggerson — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/EditCardServlet.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import java.io.IOException;
+
+import com.google.gson.Gson;
+import com.google.sps.tool.ResponseHandler;
+import com.google.sps.data.Folder;
+import com.google.sps.tool.GoogleTranslationAPI;
+import java.util.Map;
+import java.util.List;
+
+@WebServlet("/editcard")
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/EditCardServlet.java:28]: Same for this PR; please use other PR comments to infer changes, and resolve this when you feel that you have finished.

File: src/main/java/com/google/sps/servlets/EditCardServlet.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      String newToLang = request.getParameter("toLang");
+      String cardKey = request.getParameter("cardKey");
+      String newRawText = request.getParameter("rawText");
+      String newTextTranslated = GoogleTranslationAPI.translateText(newRawText, newToLang);
+      String newBlobKey = getBlobKey(request);
+      updateCard(response, newLabels, newfromLang, newToLang, cardKey, newRawText, newTextTranslated, newBlobKey);
+    }
+  }
+
+  private void updateCard(
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/EditCardServlet.java:59]: This function and the `setCardWithRequestedEdits` function duplicate logic in Card. Please fix.

File: src/test/java/com/google/sps/servlets/TranslationServletTest.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    helper.setEnvIsLoggedIn(true);
+    helper.tearDown();
+  }
+
+  @Test
+  public void TranslateHelloToSpanish() throws Exception {
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/TranslationServletTest.java]: Naming.

File: src/test/java/com/google/sps/servlets/EditCardServletTest.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    helper.tearDown();
+  }
+
+  @Test
+  public void EditCard() throws Exception {
```
Comment (tylerhou) [inline src/test/java/com/google/sps/servlets/EditCardServletTest.java]: naming.

File: src/main/java/com/google/sps/servlets/EditCardServlet.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.List;
+
+@WebServlet("/editcard")
+public class EditCardServlet extends HttpServlet {
+
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (userService.isUserLoggedIn()) {
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/EditCardServlet.java]: Avoid indentation when possible. I would prefer to write this as:

```js
if (!userService.isUserLoggedIn()) {
  // send error ... ?
  return;
}

String cardKey = request.getParameter("cardKey");
...
```

File: src/main/java/com/google/sps/servlets/EditCardServlet.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  private String getImageBlobKey(HttpServletRequest request) {
+    // Method to determine whether or not this is a unit test or live server
+    // Unit tests will always set blobKey to "null"
+    // There should be no paramater testStatus in the live server thus returns null
+    String blobKey;
+
+    if (request.getParameter("testStatus") == null) {
+      blobKey = getBlobKeyFromBlobstore(request, "image");
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/EditCardServlet.java]: return directly here

File: src/main/java/com/google/sps/servlets/TranslationServlet.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import com.google.gson.Gson;
+
+@WebServlet("/translation")
+public class TranslationServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (userService.isUserLoggedIn()) {
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/TranslationServlet.java]: Same here; you should return early if not logged in.

**Review by tylerhou — APPROVED** [review]

**Review by zmr — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/EditCardServlet.java (hunk 8 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/EditCardServlet.java]: done

File: src/main/java/com/google/sps/servlets/EditCardServlet.java (hunk 7 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/EditCardServlet.java]: done

**Pull request merged**